### ISSUE TITLE
Add a new "base64url" option for `ser_json_bytes`

### ISSUE THREAD

**Issue by MasterKale** [pydantic/pydantic#7000]
### Initial Checks

- [X] I have searched Google & GitHub for similar requests and couldn't find anything
- [X] I have read and followed [the docs](https://docs.pydantic.dev) and still think this feature is missing

### Description

In some projects, like [py_webauthn](https://github.com/duo-labs/py_webauthn), it is desirable to be as explicit as possible about the choice of encoding for `bytes <-> str` serialization for sake of a clearer developer experience.

In my case I specifically use base64url everywhere because [the WebAuthn spec explicitly calls it out as a dependency](https://www.w3.org/TR/webauthn-2/#sctn-dependencies). Thus I opt for exclusive use of `base64.urlsafe_b64decode` and `base64.urlsafe_b64encode`.

With Pydantic v1 I could accomplish this with custom `json_encoders` and `json_loads`:

```py
class WebAuthnBaseModel(BaseModel):
    """
    A subclass of Pydantic's BaseModel that includes convenient defaults
    when working with WebAuthn data structures

    `modelInstance.json()` (to JSON):
    - Encodes bytes to Base64URL
    - Converts snake_case properties to camelCase

    `Model.parse_raw()` (from JSON):
    - Decodes Base64URL to bytes
    - Converts camelCase properties to snake_case
    """

    class Config:
        json_encoders = {bytes: bytes_to_base64url}
        json_loads = json_loads_base64url_to_bytes
        alias_generator = snake_case_to_camel_case
        allow_population_by_field_name = True

    ...
```

Would you consider adding a new "`base64url`" option for [`ConfigDict.ser_json_bytes`](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.ser_json_bytes) that would use the `urlsafe_` variants of Python's base64 decoding and encoding, to ensure that base64url characters are used _and_ ensure that trailing `=` are stripped out as well?

Example use case:

```py
from pydantic import BaseModel, ConfigDict


class FooModel(BaseModel):
    model_config = ConfigDict(
        populate_by_name=True,
        ser_json_bytes='base64url',  # Proposed addition
    )

    field_1: bytes
    field_2: bytes


foo_instance = FooModel(
    field_1=bytes([1, 2, 3, 4, 5]),
    field_2=bytes([1, 2, 3, 4, 5]),
)

print(foo_instance.model_dump_json())
# Desired output: {"field_1":"AQIDBAU","field_2":"AQIDBAU"}
```

The same `FooModel` using the existing `base64` outputs the following:

```json
{"field_1":"AQIDBAU=","field_2":"AQIDBAU="}
```

It appears I might be able to pull this off with Pydantic v2 as-is if I liberally sprinkle `@field_serializer`'s everywhere, but things could look much cleaner if `ser_json_bytes` gained a "`base64url`" option instead.

Or perhaps some way to specify a custom serializer to use for serializing bytes fields instead as a new `ConfigDict` argument?

Thank you for considering this all of this.

### Affected Components

- [ ] [Compatibility between releases](https://docs.pydantic.dev/changelog/)
- [ ] [Data validation/parsing](https://docs.pydantic.dev/usage/models/#basic-model-usage)
- [X] [Data serialization](https://docs.pydantic.dev/usage/exporting_models/) - `.model_dump()` and `.model_dump_json()`
- [ ] [JSON Schema](https://docs.pydantic.dev/usage/schema/)
- [ ] [Dataclasses](https://docs.pydantic.dev/usage/dataclasses/)
- [X] [Model Config](https://docs.pydantic.dev/usage/model_config/)
- [ ] [Field Types](https://docs.pydantic.dev/usage/types/) - adding or changing a particular data type
- [ ] [Function validation decorator](https://docs.pydantic.dev/usage/validation_decorator/)
- [ ] [Generic Models](https://docs.pydantic.dev/usage/models/#generic-models)
- [ ] [Other Model behaviour](https://docs.pydantic.dev/usage/models/) - `model_construct()`, pickling, private attributes, ORM mode
- [ ] [Plugins](https://docs.pydantic.dev/) and integration with other tools - mypy, FastAPI, python-devtools, Hypothesis, VS Code, PyCharm, etc.

Selected Assignee: @samuelcolvin

**Comment by samuelcolvin** [comment]
you can do this now using `Annotated` to create your own custom type that is a thin wrapper around `bytes`:

```py
import base64
from typing import Annotated

from pydantic import BaseModel, PlainSerializer, PlainValidator
from devtools import debug


UrlSafeBase64Bytes = Annotated[
    bytes,
    PlainValidator(lambda x: base64.urlsafe_b64decode(x)),
    PlainSerializer(lambda x: base64.urlsafe_b64encode(x), return_type=bytes, when_used='json')
]


class FooModel(BaseModel):
    field: UrlSafeBase64Bytes

f = FooModel(field=base64.b64encode(b'fish pie'))
debug(f)
debug(f.model_dump(mode='json'))
debug(f.model_dump_json())
```

In future we'll provide a way to customise the core schema (and thereby validation and serialization logic) for all fields of a given type.

I'd also be happy to accept a PR to add `base64url` support during serialization to pydantic-core - therefore keeping this open.

**Comment by ljodal** [comment]
Hey and thanks for the reply @samuelcolvin! It is indeed possible to do that using a custom type, and I've done just that in [a PR for the webauthn project](https://github.com/duo-labs/py_webauthn/pull/163) mentioned here. The problem with that approach, as far as I can tell, is that it's impossible to support both v1 and v2 of pydantic at the same time. It also requires changing all downstream usages of bytes fields, whereas the old solution of using the `json_encoders`/`json_loads` options allowed for configuring this in a base model which would then have this logic applied automatically to all subclasses.

This probably isn't a very wide use case and I managed to make it work by utilizing a custom model serializer to post-process the encoded base64 strings, but it's not very pretty and probably not performant either:

https://github.com/duo-labs/py_webauthn/blob/ecf34463493da4941c7715d73a9247f9f93dcf09/webauthn/helpers/structs.py#L84-L99

**Comment by Viicos** [comment]
I've just faced an issue while working with `ser_json_bytes='base64'`, and it seems `pydantic-core` is already using [URL safe encoding](https://github.com/pydantic/pydantic-core/blob/e22a2fd03ef86bb33972bfa4d90cd8e3268898c9/src/serializers/config.rs#L150-L157)?

Noticed this while trying to do the following:

```python
import os
from base64 import b64decode, urlsafe_b64decode

from pydantic import BaseModel

class Model(BaseModel):
    a: bytes

    model_config = {
        "ser_json_bytes": "base64",
    }

model = Model(a=os.urandom(16))
print(model.model_dump_json())
#> {"a":"_xHDtXO6BTYZuCccN1bgeQ=="}
b64decode("_xHDtXO6BTYZuCccN1bgeQ==")  # raises error
urlsafe_b64decode("_xHDtXO6BTYZuCccN1bgeQ==")
#> b"\xff\x11\xc3\xb5s\xba\x056\x19\xb8'\x1c7V\xe0y"
```

By the way, according to [RFC4648](https://datatracker.ietf.org/doc/html/rfc4648#section-5), the `=` padding isn't necessarily removed, it can be omitted or percent-encoded

**Comment by samuelcolvin** [comment]
Thanks @Viicos - great point - it seems like the base64 encoding using by pydantic-core already matches `urlsafe_b64encode`.

Closing this, but @MasterKale let us know if you have any other requests - would be great to get py_webauthn using pydantic V2.

**Comment by MasterKale** [comment]
In light of the fact that `ser_json_bytes="base64"` does in fact use base64url encoding (@samuelcolvin perhaps [the docs here](https://docs.pydantic.dev/latest/api/config/#pydantic.config.ConfigDict.ser_json_bytes) can get an update to reflect this? It might trip people up who try to base64-decode a base64url-encoded value because they see `"base64"`), then there's one last thing I want to confirm.

In my original feature request I noted that I'm interested in stripping out the padding:

```py
print(foo_instance.model_dump_json())
# Desired output: {"field_1":"AQIDBAU","field_2":"AQIDBAU"}
```

One way you mentioned is to use `Annotated` as shown above; is @ljodal's suggestion also a "correct" (in terms of performance, I suppose) way to accomplish this, or is it an unintended use of `@model_serializer()`?

```py
@model_serializer(mode="wrap", when_used="json")
def clean_up_base64(
    self, serializer: Callable[..., Dict[str, Any]]
) -> Dict[str, Any]:
    """
    Remove trailing "=" from bytes fields serialized as base64 encoded strings.
    """

    serialized = serializer(self)

    for name, field_info in self.model_fields.items():  # type: ignore[attr-defined]
        value = serialized.get(name)
        if field_info.annotation is bytes and isinstance(value, str):
            serialized[name] = value.rstrip("=")

    return serialized
```

In an earlier attempt to use `Annotated` I thought too much of a model's readability was lost by swapping `bytes` type annotations for something like `UrlSafeBase64Bytes` (when I see a value like that it kinda reads like there's something about the bytes that mean it's tied to a specific encoding, when you and I both know it's just `bytes`), thus my desire to find a way to strip the padding without `Annotated`.

**Comment by samuelcolvin** [comment]
Thanks @MasterKale, @Viicos has kindly updated the documentation.

Both `Annotated[..., WrapSerializer]` and `field_serializser` are equally correct. `model_serializser` will also work, but will have some performance overhead.

---

Long term, we have a plan to allow schema generation to be customised, this work has started in #6737 in response to #6045, but it turns out it's f**king complicated (you can see three PRs attempting to solve #6045) and getting it write requires some more fundamental changes to pydantic-core schema.

TL;DR in time we'll provide you with a way to customise the serialization behaviour of all all types.

cc @adriangb if we create an issue to track your work on making `GenerateSchema` public, we should link to this use case.
